include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort_one_wg.h

```diff
@@ -173,10 +173,23 @@ struct __subgroup_radix_sort
                         uint16_t __begin_bit = 0;
                         constexpr uint16_t __end_bit = sizeof(_KeyT) * ::std::numeric_limits<unsigned char>::digits;
 
+                        auto __val_mem_adjusted_barrier = [__it]() {
+                            if constexpr (_SLM_tag_val::value)
+                                __dpl_sycl::__group_barrier(__it);
+                            else
+                                __dpl_sycl::__group_barrier(__it, __dpl_sycl::__fence_space_global_and_local{});
+                        };
+                        auto __count_mem_adjusted_barrier = [__it]() {
+                            if constexpr (_SLM_counter::value)
+                                __dpl_sycl::__group_barrier(__it);
+                            else
+                                __dpl_sycl::__group_barrier(__it, __dpl_sycl::__fence_space_global_and_local{});
+                        };
+
```

Comment:
This just a thought, but could we use a single lambda here and pass the boolean constant as a parameter to it? Something like:
```c++
    auto __mem_adjusted_barrier = [__it](auto __is_slm) {
        if constexpr (decltype(__is_slm)::value)
            __dpl_sycl::__group_barrier(__it);
        else
            __dpl_sycl::__group_barrier(__it, __dpl_sycl::__fence_space_global_and_local{});
    };
```
And when used:
```c++
   __mem_adjusted_barrier(_SLM_tag_val{});
```
```c++   
  __mem_adjusted_barrier(_SLM_counter{});
```

See https://godbolt.org/z/GMx64dTb5.

Reply:
Why do we need a compile time tag here?
Whether is it simpler to call with a constant, wrapped into `__dpl_sycl::name_xx`, like?

`__dpl_sycl::__group_barrier(__it, __dpl_sycl::fence_global_and_local);`

Reply:
@MikeDvorskiy, I do not understand how you suggesting is supposed to work. We need to use different fence arguments. Could you be more specific? Perhaps I just do not get your idea.

Reply:
Meanwhile, I've implemented what Adam suggested.